### PR TITLE
MM-43143: Dataloaders for roles

### PR DESCRIPTION
**Pull request by agnivade**
We use a dataloader to optimize roles loading.

https://mattermost.atlassian.net/browse/MM-43143

```release-note
NONE
```

### CODE REVIEW

File: api4/resolver_test.go [hunk 1]
```diff
@@ -143,3 +143,85 @@ func TestGraphQLChannelsLeft(t *testing.T) {
...
+	}
+	`,
+	}
+
+	resp, err := th.MakeGraphQLRequest(&input)
+	require.NoError(t, err)
+	require.Len(t, resp.Errors, 0)
+	require.NoError(t, json.Unmarshal(resp.Data, &q))
+
+	assert.Len(t, q.User.Roles, 1)
```
Comment (isacikgoz) [inline api4/resolver_test.go:213]: I think we shouldn't continue if we have let's say `0` elements in the slice.
```suggestion
	require.Len(t, q.User.Roles, 1)
```

**Review by isacikgoz — COMMENTED** [review]
Looks good, just some small improvements needed to avoid panics in tests.

File: api4/resolver_channel_member.go [hunk 2]
```diff
@@ -65,12 +66,24 @@ func (cm *channelMember) Channel(ctx context.Context) (*channel, error) {
...
+	thunk := loader.LoadMany(ctx, dataloader.NewKeysFromStrings(strings.Fields(cm.Roles)))
+	results, errs := thunk()
+	// All errors are the same. We just return the first one.
+	if len(errs) > 0 && errs[0] != nil {
+		return nil, err
+	}
+
+	roles := make([]*model.Role, len(results))
+	for i, res := range results {
+		roles[i] = res.(*model.Role)
```
Comment (agnivade) [inline api4/resolver_channel_member.go:83]: Btw, I think dataloaders are a prime candidate to use generics. The whole thunk API uses `interface{}`. I have talked with the maintainers and they agree as well. Let's see..

Comment (isacikgoz) [reply api4/resolver_channel_member.go:83]: Cool, looking forward to see proper uses of generics :)

**Review by isacikgoz — APPROVED** [review]
Nice!

**Pull request merged**